task/net/src/vlan.rs

```diff
@@ -0,0 +1,407 @@
+            let mut iface = builder
+                .hardware_addr(mac.into())
+                .neighbor_cache(neighbor_cache)
+                .ip_addrs(ipv6_net)
+                .finalize();
+
+            // Associate sockets with this interface.
+            let socket_handles = socket_handles_iter.next().unwrap();
+            assert_eq!(sockets.len(), SOCKET_COUNT);
+            for (s, h) in sockets.into_iter().zip(&mut socket_handles[..]) {
+                *h = iface.add_socket(s);
+            }
+            // Bind sockets to their ports.
+            assert_eq!(socket_handles.len(), SOCKET_COUNT);
+            assert_eq!(generated::SOCKET_PORTS.len(), SOCKET_COUNT);
+            for (&h, &port) in
+                socket_handles.iter().zip(&generated::SOCKET_PORTS)
+            {
+                iface
+                    .get_socket::<UdpSocket>(h)
+                    .bind((ipv6_addr, port))
+                    .map_err(|_| ())
+                    .unwrap();
+            }
+            *ifaces_iter.next().unwrap() = Some(iface);
+
+            // Increment the MAC and IP addresses so that each VLAN has
+            // a unique address.
+            ipv6_addr.0[15] += 1;
+            mac.0[5] += 1;
+        }
+
+        let ifaces = ifaces.map(|e| e.unwrap());
+        Self {
+            eth: &storage.eth,
+            socket_handles,
+            ifaces,
+            bsp,
+        }
+    }
+
+    pub fn poll(&mut self, t: u64) -> smoltcp::Result<bool> {
+        let t = smoltcp::time::Instant::from_millis(t as i64);
+        let mut any_activity = false;
+        for iface in &mut self.ifaces {
+            any_activity |= iface.poll(t)?;
+        }
+        Ok(any_activity)
+    }
+
+    /// Iterate over sockets, waking any that can do work.  A task can do work
+    /// if all of the (internal) VLAN sockets can receive a packet, since
+    /// we don't know which VLAN it will write to.
+    pub fn wake_sockets(&mut self) {
+        for i in 0..SOCKET_COUNT {
+            if (0..VLAN_COUNT)
+                .any(|v| self.get_socket_mut(i, v).unwrap().can_recv())
+            {
+                let (task_id, notification) = generated::SOCKET_OWNERS[i];
+                let task_id = sys_refresh_task_id(task_id);
+                sys_post(task_id, notification);
+            }
+        }
+    }
+
+    pub fn wake(&self) {
+        self.bsp.wake(&self.eth)
+    }
+
+    fn get_handle(
+        &self,
+        index: usize,
+        vlan_index: usize,
+    ) -> Result<SocketHandle, RequestError<NetError>> {
+        self.socket_handles
+            .get(vlan_index)
+            .ok_or(RequestError::Fail(ClientError::BadMessageContents))
+            .and_then(|s| {
+                s.get(index)
+                    .cloned()
+                    .ok_or(RequestError::Fail(ClientError::BadMessageContents))
+            })
+    }
+
+    /// Gets the socket `index`. If `index` is out of range, returns
+    /// `BadMessage`.
```

Comment:
Is it worth noting that an out of bounds `vlan_index` will panic? I wouldn't normally suggest that since this is a private function, but since the docstring explicitly states what happens with an out of bounds `index` it might be worth it.

Reply:
Sure, added a comment